### PR TITLE
Support `phx-change` on all elements with `@LiveBinding`

### PR DESCRIPTION
**Pull request by carson-katri**
Depends on https://github.com/liveview-native/live_view_native_platform/pull/10

You can now use `phx-change` in place of `native_binding` to manually update the values. This was implemented already for form elements (which use `@FormState`), but is now possible for other types of element, such as `List`.

On form elements, the value sent will be in the format `%{ "value" => new_value }`. If you set a `name` attribute, it will become `%{ "your_custom_name" => new_value }`.

Other types of elements will use the name of the attribute (such as, `%{ "selection" => new_value }`). This allows multiple attributes to send change events without collisions.

Here's an example converting code that uses `native_binding` to use `phx-change`.

<table>
<tr>
<td>

```elixir
defmodule TestBedWeb.IndexLive do
  use TestBedWeb, :live_view
  use LiveViewNative.LiveView

  native_binding :selection, :string, default: "1"

  def render(assigns) do
    ~SWIFTUI"""
    <List selection="selection">
      <Text :for={item <- 1..10} id={"#{item}"}>
        <%= item %>
      </Text>
    </List>
    """
  end
end
```

</td>
<td>

```elixir
defmodule TestBedWeb.IndexLive do
  use TestBedWeb, :live_view
  use LiveViewNative.LiveView

  def mount(_params, _session, socket) do
    {:ok, assign(socket, selection: "1")}
  end

  def render(assigns) do
    ~SWIFTUI"""
    <List selection={@selection} phx-change="on_select">
      <Text :for={item <- 1..10} id={"#{item}"}>
        <%= item %>
      </Text>
    </List>
    """
  end

  def handle_event("on_select", %{ "selection" => new_value }, socket) do
    {:noreply, assign(socket, selection: new_value)}
  end
end
```

</td>
</tr>
</table>

### CODE REVIEW

**Comment by bcardarella** [comment]
So I want to use this PR as a place to discuss this change. It is a bit of a fork in the road for us and let's lay out all of the pros/cons.

For some background, we've had a `value_binding` feature in LVN for a while now that allows for a two-way data binding between the server state and the client state. When a value is updated on the client the binding works to update the server state and when the server updates the value the client is updated.

## value_binding

### Pros

It is far easier/quicker to use and it has been well tested to ensure we're encapsulating all of the needs of SwiftUI. SwiftUI is *not* a web client so we should expect that solutions developed in LiveView itself were done in the context of what makes sense for the web whereas we are tasked with what makes sense for SwiftUI app and also what makes sense for the other native clients that we hope to support in the future.

### Cons

We are deviating away from the expected way that LiveView web works. In that world any state changes on the client that need to be reflected on the server are opt-in through events. In some cases the `phx-change` event along with a named event and the values that are being passed back as arguments. In this we'd align more closely with the LiveView API contract. Because `value_binding` works for SwiftUI and makes sense for it doesn't guarantee that it will make sense for other native clients. So we're opening the door to other clients offering their own two way data binding solution. Now this creates divergent API that complicates the learning process.

## phx-change

### Pros

If we do move toward the `phx-change` side of things and align more closely with LiveView's own implementation we should make every effort to surface all of the edge cases on where `value_binding` is solving for us that may be complicated or possibly impossible to solve with the change event handling.  We are also creating a more implicit contract that devlopers are required to implement their own `handle_event` functions. This gives developers more insight into exactly what is happening under the hood.

### Cons

Our team has been happily using `value_binding` for months and has reported that it works and works well. Deviating from this is unfortunately turning back the clock on a lot of work that went into making this feature work flawlessly.

**Comment by bcardarella** [comment]
A few other concerns with the current implementation:

1. Ecto is a dependency that teams are forced into using. What if they don't want Ecto?
2. The `handle_event('_native_bindings', ...)` function doesn't give fine grained control to the developer on how to handle the data once it gets to the server.
3. there is nothing actually "native" here as this is a pattern that could easily also exist on the web. However, after a few years of LiveView itself being used in production by many this pattern has not emerged. We are pushing something that hasn't naturally come from use cases

**Comment by carson-katri** [comment]
One thing that still needs to be solved is how change events are declared for modifiers. Native bindings are used for quite a few modifiers:

```elixir
sheet(is_presented: :my_presented_binding)
```

One option would be to send a map with `value` and `change` keys instead:

```elixir
sheet(is_presented: %{ value: @is_presented, change: "presentation_changed" })
```

Alternatively, we could add an extra key to the modifier for the change event.

```elixir
sheet(is_presented: @is_presented, change: "presentation_changed")
```

However, this would limit us to a single event for all bindings on a modifier, whereas the first approach allows for a separate event on each.

**Comment by bcardarella** [comment]
@carson-katri this is exactly the type of edge cases I want to continue to surface. If there are any more things that `native_binding` was solving for us that is now not so easy (or hopefully not "impossible") to solve with this change let's vet it all out

**Comment by carson-katri** [comment]
One other thing to consider:
From what I can tell, LV does not support `phx-change` outside of the context of a `form`.

Would we need to come up with our own name for this attribute, such as `lvn-change`? Or is this an acceptable extension of `phx-change`'s behavior?

**Comment by bcardarella** [comment]
@carson-katri I believe the `change` event applies to all eventListeners that support that JS event which I think are all inputs: https://github.com/phoenixframework/phoenix_live_view/blob/9ea1402ce162abac90a1b7aff82e7bd528528e51/assets/js/phoenix_live_view/view.js#L895

**Comment by bcardarella** [comment]
To articulate my concerns, managing multiple sources of truth can lead to concurrency issues. If the purpose of `persist` doesn't create this problem I'm OK with it but if we're just using it for token storage maybe we can find another way to do it?

**Comment by kgautreaux** [comment]
Disclaimer: I have not used LVN in anger yet. My use case relies on Android support so I was waiting for ElixirConf before tracking the main branch and working on an MVP.

Given that I don't have intimate knowledge of the current design goals I am more inclined to favor the `phx-change` approaching than the `native_binding` approach because I suspect that:

1.  It will align better with my current LV knowledge (particularly around Streams and [form recovery](https://hexdocs.pm/phoenix_live_view/form-bindings.html#submitting-the-form-action-over-http))
2. Since I am not a SwiftUI or JetPack Compose expert having less magic in the event handling of the LV code will help me reason about my data-flow and recognize when I need to dig into those frameworks to debug when the behavior deviates from the web UI.
3. It will allow better code reuse and push deviations down strictly into the presentation layer if `web`, `swiftui`, and `android` all use similar data flow.
4. I don't care about the ecto dependency because all apps will be cross-platform and need a persistence layer. (I'm not going to use CloudKit for the Web)

Philosophically, I think a strong case can be made that the target audience for this product is not a React Native, Flutter, or Android/iOS consultancy looking for a better cross-platform development story, nor is it an Android or iOS shop looking for a more efficient way to develop their existing native apps, but it is the existing Elixir dev or team looking to add native mobile app support to their SaaS product without hiring a native dev or contracting that work out.  With that audience top of mind, aligning with their current LV practices holds more sway than decreasing the lines of code.

If we are extending existing web LV products onto native then we have to write the camera, location, HealthKit, HomeKit, WidgetKit, VisionOS etc...code de novo anyway, so it makes it much easier to just rewrite the presentation layer code for our existing LV CRUD form-heavy LiveViews and then add on additional modules to handle camera, etc...that we need for mobile, perhaps even using platform specific code.

Lastly, if SwiftUI and Jetpack Compose gain their benefit from avoiding state and being declarative then I don't think we should obfuscate the 'Actions Up' phase of declarative UI interaction. However, this is a DY product, not my product, so make the decision that will increase adoption and generate contract engagement.

**Comment by bcardarella** [comment]
@kgautreaux the streams argument is a very good one and I've confirmed with Jose that this very well could be a blocker for the `native_binding` direction all together. We definitely want to support streams, that work is currently happening in Core right now to get streams supported in our channels client.

**Comment by stevegraham** [comment]
I don't have a strong opinion on this immediately but some thoughts that crossed my mind while reading?
 
- Why the change? Especially if DY has been busy cranking with the existing pattern seemingly without issue?
- I'm not persuaded that the pattern not emerging on the web is dispositive as to its merits for native GUI apps
- There is a lot of boiler plate with the proposal, which is kind of the point of two-way data bindings.
- Why is this a choice, why must the current pattern be totally thrown out in favor of the new?

Maybe I'm missing the point (likely) but I can't see a need for this change from what you've written. I don't see what problem is being fixed?

That being said I defer to @bcardarella and @carson-katri on this. The above is just what ran through my mind reading this thread.

**Comment by bcardarella** [comment]
@stevegraham I think regardless of the other reasons once I heard that Streams can't be supported by `native_binding` that's a no-go for me.

**Comment by bcardarella** [comment]
> One thing that still needs to be solved is how change events are declared for modifiers. Native bindings are used for quite a few modifiers:
> 
> ```elixir
> sheet(is_presented: :my_presented_binding)
> ```
> 
> One option would be to send a map with `value` and `change` keys instead:
> 
> ```elixir
> sheet(is_presented: %{ value: @is_presented, change: "presentation_changed" })
> ```
> 
> Alternatively, we could add an extra key to the modifier for the change event.
> 
> ```elixir
> sheet(is_presented: @is_presented, change: "presentation_changed")
> ```
> 
> However, this would limit us to a single event for all bindings on a modifier, whereas the first approach allows for a separate event on each.

@carson-katri can you roll back that commit you made to address ☝️ let's spend some time disecting the problem so we can ensure we're putting together the best API

**Comment by stevegraham** [comment]
> @stevegraham I think regardless of the other reasons once I heard that Streams can't be supported by native_binding that's a no-go for me.

Do I understand this correctly: it is not possible to use streams if native_binding is used in the same liveview for anything, even something completely unrelated?

**Comment by stevegraham** [comment]
Seems like you could keep the `native_binding` API if its implementation was a macro that generated code semantically equivalent to the code on the right, i.e. enclose `mount` in a function that sets the initial "binding" value, generates an "update_BINDINGNAME" event handler, and the underlying view implementations knows to use phx-change to update the "binding". You could have the exact same API, no?

**Comment by bcardarella** [comment]
Having two difference ways to do something is a big red flag for me.

**Comment by mayel** [comment]
I support the new approach with `handle_event` for re-usability of code between web and native 👍

**Comment by stevegraham** [comment]
It's not two ways to do it. It's the same way.

**Comment by stevegraham** [comment]
If you've made your mind up IMO you should close this issue to be respectful of other people's time. 

I see a way to keep the API as-is whilst implementing it to generate semantically equivalent code to the right in @carson-katri's example, but it doesn't need to be in LVN core. It could be implemented by anyone and published as a hex library.

The only thing I would suggest that would make the library solution nicer is keeping the SWIFTUI hex template API the same, i.e.

```
<List selection="selection">
  <Text :for={item <- 1..10} id={"#{item}"}>
    <%= item %>
  </Text>
</List>
```

`selection="selection"` means it's concerned with an assign called `@selection` and assumes there is an event handler `update_selection` that updates `@selection` implemented in the LV. You might think this this is too magical, that's ultimately a matter of taste and your decision to make.


> Having two different ways to do something is a big red flag for me. And we certainly shouldn't be promoting API that is dead ended.

Disagree with this. Many counterexamples in Elixir.

**Comment by bcardarella** [comment]
@stevegraham we're still soliciting feedback. Right now I haven't heard an objective reason on the `native_binding` side. We have the Stream limit which for me is not great. But, as far as the macro goes there's no reason why the macro cannot be extracted out and shared for those that want to use it.

**Comment by stevegraham** [comment]
The objective reason is it's a standard pattern in GUI applications, and the alternative solution (as proposed) results in objectively messier application code. 

The only legitimate reason against `native_binding` that has been discussed here is the current implementation purportedly does not work with streams (although it hasn't been explained how?). However, it seems possible to reimplement `native_binding` such that it becomes compatible, because it seems possible to implement it with a bit of convention over configuration and a macro emitting code semantically equivalent to the new solution at compile-time.

FWIW I'm not sure how useful some of the views on this thread actually are. All this talk about code re-use, but how many have actually written any LVN code, or spent significant time playing with it.

**Comment by stevegraham** [comment]
This:

> However, after a few years of LiveView itself being used in production by many this pattern has not emerged. We are pushing something that hasn't naturally come from use cases

is explained by this:

> We are deviating away from the expected way that LiveView web works. In that world any state changes on the client that need to be reflected on the server are opt-in through events.

re this:

> Because value_binding works for SwiftUI and makes sense for it doesn't guarantee that it will make sense for other native clients.

Should be pretty straightforward to look at how the targeted platforms solve for this?

**Comment by bcardarella** [comment]
@stevegraham does the extraction of that macro satisfy your concerns here?

**Comment by drewchandler** [comment]
I think `Picker` might need some updates. If I change an assign on the server-side and use the `selection` attribute, the UI is not updated. If I use `value` instead, it is. Should the attribute name match the naming used by the [swift component](https://developer.apple.com/documentation/swiftui/picker)?

**Comment by bcardarella** [comment]
@drewchandler I think that would be a separate issue, can you open that? It should match the SwiftUI api in Apple's docs. So in this branch for testing we'd want:

```
<Picker selection={@selection} phx-change="picker_change" />
```

**Comment by josevalim** [comment]
Folks, I am jumping into the thread a bit late, but does anyone have production examples of `native_binding` being used?

The reason I ask is the following: in all my time using LiveView, I can't think of a single time where a value was sent to the server only to be stored as an assign and nothing else. Every time a value is sent to the server, it usually triggers other actions, and it may or may not set an assign of said name. Therefore, I am wondering under which circumstances this would be necessary for LVN.

For this reason, I am worried this pattern can lead to misuse. For example, imagine I have a category selector, which sends a "category_id" to the server via `native_binding`. Also a category may have subcategories, which are loaded based on the "category_id". Because `native_binding` automatically sets `@category_id`, some may naively push the logic that loads all subcategories to the template because that's the only place they can react to the `@category_id` change.

Two additional nitpicks/questions: I don't think there is anything in `native_binding` that actually ties it to `native` is there? If this is proven to be a useful pattern based on examples, it may be worth changing its name to something generally applicable. Also, it seems it uses Ecto for casting, and it seems like a bit of an overkill to bring a large library such as Ecto with the only goal of casting assigns.

Thank you for inviting the discussion and have a great weekend :heart:

**Comment by mayel** [comment]
> in all my time using LiveView, I can't think of a single time where a value was sent to the server only to be stored as an assign and nothing else. Every time a value is sent to the server, it usually triggers other actions

This is mostly my experience as well, but sometimes the action triggered may be in a child component, and all that happens in the actual view or live_component where the event was triggered in that case is setting an assign. Here's a recent example:
- `ThreadLive` contains a component `.filters` and a live component `CommentsLive`
- `.filters` has UI which trigger events (eg. `sort_by`) which are sent to `ThreadLive`
- `ThreadLive` simply sets assigns (eg. `sort_by`) which are passed to:
  - `.filters` for the UI to reflect the current filters
  - `CommentsLive` to handle the loading/filtering of comments (that code is there rather than `ThreadLive` because that component is used in other places like `PostLive` etc)

**Comment by bcardarella** [comment]
@josevalim we don't yet have a production app of significant complexity in production. the LV patterns are pretty well battle tested at this point and with our limited budget I'm always looking for pre-existing works to leverage to reduce our overhead. At it's most simple LVN is essentially an intermediary layer between the already built, working, and proven LiveView programming model and the already built, working, and proven native clients. This debate I find comes down to do you want the app building experience to be more LiveView-esque or more SwiftUI-esque. While I realize, and have witnessed, trying to get SwiftUI things working on server and LV can at times feel very square peg - round hole I think eventually we've come around to mostly good compromises. In this particular case I bear responsibility for letting `native_binding` hang around this long. Honestly it was one of those tunnel vision situations where I was focused on higher mind things and through some internal app building last week I realized we still had this implemented. It kicked off an internal discussion on its need and if we could and should deviate from it. I understand that taking things away sometimes sucks and I've had my fair share of time being on the receiving end of that in OSS projects but ultimately I'm looking at what is going to enable us long-term for success and what is going to be best for Elixir devs to build native apps. So I tend to side on the `phx-change` side of the argument before the whole Stream issue came about. Elixir/Phoenix/LiveView devs are our intended consumers of LVN so I'm of the mind that we keep the development experience as familiar to LV as much as possible. The technical limits I think are the nail in the coffin. We could very well expand `native_binding` but now that's a new API to support and that more surface area. I think if we extract as a macro for those that want to continue to use it that's OK but the recommended path will be what this PR pushes us towards.

**Comment by josevalim** [comment]
> @josevalim we don't yet have an app of significant complexity in production

Sure, I expected that, but even a demo with actual functionality suffices. It is only because all the examples I have seen so far that used `native_binding` was in the examples itself, where the LV has nothing to do.

> In this particular case I bear responsibility for letting native_binding hang around this long. Honestly it was one of those tunnel vision situations where I was focused on higher mind things and through some internal app building last week I realized we still had this implemented. 

Nah, don't worry too much about it. This is completely normal and expected, _especially_ as you have larger concerns. There are several examples of features that existed in Elixir and Ecto which turned out to not be a good idea, but we ironed them out towards 1.0 (in Ecto's case, towards 2.0 :P).

Something else I will add: removing a shipped feature is hard, adding a non-released feature is easy. So if you are unsure, remove now, consider use cases, and add later.

**Comment by bcardarella** [comment]
> Something else I will add: removing a shipped feature is hard, adding a non-released feature is easy. So if you are unsure, remove now, consider use cases, and add later.

Exactly where my ehad is at

**Comment by drewchandler** [comment]
I am very torn on `native_binding` vs `phx-change`. My app is basically a giant form and with `phx-change` I am having to write a lot of boilerplate `handle_event` functions to handle every field. I do like the explicitness of `phx-change` and it fits with the “data down, actions up” style I am used to. Without `phx-change` , would I have use modifiers or something if I wanted to validate a field?

There are 2 things from the web that I find myself missing:
1. On the web you can just do `phx-change` on the `<form>` instead of having to do every input. In SwiftUI, the `Form` component seems like it is just a presentational thing and doesn’t have any events.
2. `phx-change` events on html elements give you a `_target` in the params, so you can tell what element triggered the event. If I had the target, I could maybe try writing a single `handle_event` and use the target to know what field to update.

I guess I have been trying to write LVN code exactly how I would write something in LV, and there is only so far that can take me, because the two platforms are different and will require different patterns.

**Comment by bcardarella** [comment]
> * On the web you can just do `phx-change` on the `<form>` instead of having to do every input. In SwiftUI, the `Form` component seems like it is just a presentational thing and doesn’t have any events.
> * `phx-change` events on html elements give you a `_target` in the params, so you can tell what element triggered the event. If I had the target, I could maybe try writing a single `handle_event` and use the target to know what field to update.

On the first one we should fix `live-form` to behave this way. For the second we should add the `_target` meta data to conform with that LV js is doing: https://github.com/phoenixframework/phoenix_live_view/blob/9ea1402ce162abac90a1b7aff82e7bd528528e51/assets/js/phoenix_live_view/js.js#L55C1-L66

**Comment by josevalim** [comment]
This is a very interesting perspective I did not have @drewchandler. Basically, LiveView/HTML can bundle all inputs for you when you submit a button. So the collecting of state/params happens on the client. LiveView/Swift does not do that, so you need to store all events on the server. This could explain a need for `native_binding` indeed. So the question is if you will stick to LiveView/Swift style OR you want to provide a HTML Form-like experience within LVN by having additional form components.

**Comment by drewchandler** [comment]
I take back all the stuff I said. I did not know about [LiveForm](https://github.com/liveview-native/liveview-native-live-form).

**Comment by bcardarella** [comment]
@josevalim SwiftUI does not have concept of a "form" the same way the web does. We built https://github.com/liveview-native/liveview-native-live-form to address this. It will take all child inputs and serialize them when submitted. I believe this solves @drewchandler's issue

**Comment by stevegraham** [comment]
> @stevegraham does the extraction of that macro satisfy your concerns here?

Personally, I don't understand the desire for a more verbose solution for framework users when `native_binding` can be implemented in a streams compatible way, but to be honest I'm not really concerned about the change. The LV part is straightforward enough to implement outside of the framework as a library, or as part of an application. 

That being said I don't think this:

```
<Picker selection={@selection} phx-change="picker_change" />
```

is a better API design than what it replaces. What other type of event handler other than something updating the state of `@selection` would be concerned with a change event? I can't think of one. So why the need to explicitly define the name of the handler? Why not assume by convention that a handler called `update_THING` exists for an assign `@THING`, and that SUI uses phx-change to communicate the change to the server. It seems superfluous to make this part of the public API. This is exactly the same design principal as assuming a `native_binding` has been declared in the LV when one used in a template in the current version of LVN.

Ultimately I don't see the need to change the API of the SwiftUI layer. You can have the same API as SwiftUI, i.e.

```
<Picker selection="selection" />
```

and have that mean "send `update_selection` event with new value when `@selection` changes on the client".

I think there is a stronger argument for leaving the Swift API as-is because it's less straightforward to DRY up for framework users.

My $0.02.

**Comment by carson-katri** [comment]
I've added [`apple/swift-async-algorithms`](https://github.com/apple/swift-async-algorithms) as a dependency for the `AsyncChannel` class and `debounce`/`throttle` extensions of `AsyncSequence`. The `AsyncChannel` gives us back pressure, so we can properly await a debounced/throttled event.

The previous Combine implementation had no way of cancelling event tasks that were dropped due to a debounce or throttle, and would just leak the continuation.

File: Sources/LiveViewNative/Modifiers/Search Modifiers/FindNavigatorModifier.swift
Comment (supernintendo) [inline Sources/LiveViewNative/Modifiers/Search Modifiers/FindNavigatorModifier.swift:43]: This produces the following warning now that `container` is unused: `Initialization of immutable value 'container' was never used; consider replacing with assignment to '_' or removing it`

**Comment by supernintendo** [comment]
I think the arguments in favor of dropping the `native_bindings` abstraction outweigh those in favor of keeping it, and it seems to be a welcome change based on responses from the community and internally. I just want to mention a few other points that I haven't seen brought up here:

1. The old approach required the use of an additional macro `assign_native_bindings/3` ([link](https://github.com/liveview-native/live_view_native/pull/35/files#diff-6b5d2273b12e7bc902f4646bff44768f322a20ee20271887fcc99f481296cdf5L82)) since `assign/2` and `assign/3` are not overridable in Phoenix. Now all state updates can be handled idiomatically, simplifying both the Elixir and Swift implementations as well as end-user application code.

2. Native bindings are not composable. You can't have an arbitrary list of elements with separate native bindings since they're "baked in" at compile time via the `native_binding/3` macro. The `phx-change` pattern resolves this by allowing elements to pass dynamic values to their `handle_event/3` callbacks, allowing clientside state to be synced from elements that are generated at runtime by the server.

3. As others have pointed out, the target audience for LiveView Native is Elixir developers who are already using Phoenix and LiveView in their applications. Doing things the "LiveView way" is one of the design goals of this project; the more we can do away with unusual abstractions in the interest of familiarity, the better.

@carson-katri I merged https://github.com/liveview-native/live_view_native/pull/35 and https://github.com/liveview-native/live_view_native_platform/pull/10 so just waiting on https://github.com/liveview-native/liveview-client-swiftui/pull/1058 to be merged and then this to rebased off `main` before merging.

**Comment by stevegraham** [comment]
@supernintendo I agree with all of your points. 

For everyone on the thread let me give an example of what I mean:

New (rough) example native binding implementation:

```elixir
defmodule NativeBinding do
  defmacro native_binding([{key, value}]) do
    quote do
      @behaviour Phoenix.LiveView
      on_mount {NativeBinding, {unquote(key), unquote(value)}}

      def handle_event("update_#{unquote(key)}", %{unquote(key) => value}, socket) do
        {:noreply, assign(socket, unquote(key), value)}
      end
    end
  end

  def on_mount({key, value}, _params, _session, socket) do
    {:cont, Phoenix.LiveView.Utils.assign(socket, key, value)}
  end
end
```

Example usage:

```elixir
defmodule TestBedWeb.IndexLive do
  use TestBedWeb, :live_view
  use LiveViewNative.LiveView
  use NativeBinding

  native_binding selection: "1"

  def mount(_params, _session, socket) do
    {:ok, socket}
  end

  def render(assigns) do
    ~SWIFTUI"""
    <List selection={@selection} phx-change="update_selection">
      <Text :for={item <- 1..10} id={"#{item}"}>
        <%= item %>
      </Text>
    </List>
    """
  end
end
```

You could also have SwiftUI views send a `phx-change` of "update_#{binding_name}" with `%{binding_name => value}` by convention meaning your render function could be almost the same as is currently, i.e.

```elixir
def render(assigns) do
  ~SWIFTUI"""
  <List selection={@selection}>
    <Text :for={item <- 1..10} id={"#{item}"}>
      <%= item %>
    </Text>
  </List>
  """
end
```

That being said, users might look at the view above and think "where the hell is `@selection` coming from?", which could be avoided by just keeping the view API the same but , e.g.

```elixir
def render(assigns) do
  ~SWIFTUI"""
  <List selection="selection">
    <Text :for={item <- 1..10} id={"#{item}"}>
      <%= item %>
    </Text>
  </List>
  """
end
```

This generates semantically equivalent code to proposal, removes a bunch of boiler plate, and allows the user to manually define their assigns and event handlers when their use case requires departing from convention.

Thoughts?

**Comment by bcardarella** [comment]
@stevegraham I appreciate your interest in this issue, I believe the best course of action is to take your macro and have it as a 3rd party lib. This way you can can still benefit from that workflow and others that find benefit can as well, while those that are more accustomed to and expecting a LiveView-like experience will get that by default. We're comfortable with the `phx-change` direction and we've been given the OK by @chrismccord to extend its use beyond just `<form>` within the context of LVN.

**Comment by stevegraham** [comment]
No problem. I just wanted to make sure I was understood. IMO there are good points either way, but ultimately a direction has to be chosen and it's not my call. 🤝

**Comment by bcardarella** [comment]
@carson-katri this looks like it'll require a rebase

File: Sources/LiveViewNative/Property Wrappers/ChangeTracked.swift
Comment (carson-katri) [reply Sources/LiveViewNative/Property Wrappers/ChangeTracked.swift]: I've updated it to just use `self.value`.

**Review by supernintendo — APPROVED** [review]
🙌🏼

**Pull request merged**
